pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/MaximumPinotConsumptionDelayTracker.java

```diff
@@ -0,0 +1,306 @@
+   * this table in the current server.
+   * This function will be invoked when we receive CONSUMING -> DROPPED / OFFLINE state transitions.
+   *
+   * @param partitionId partition id that we should stop tracking.
+   */
+  public void stopTrackingPinotConsumptionDelayForPartition(long partitionId) {
+    removePartitionId(partitionId);
+  }
+
+  /*
+   * This method is used for ticking up the time a partition has been out of CONSUMING state.
+   * When that time exceeds some threshold, we read from ideal state to confirm we still host the partition, if
+   * not we remove the partition from being tracked locally.
+   * This call is to be invoked by a timer thread that will periodically wake up and invoke this function.
+   */
+  public void tickInactivePartitions() {
+    for (long partitionId : _partitionsMarkedForConfirmation.keySet()) {
+      int currentTicks = _partitionsMarkedForConfirmation.get(partitionId);
+      if (++currentTicks >= _maxNumberOfTicksBeforeVerification) {
+        if (!isPartitionHostedByThisServerPerIdealState(partitionId)) {
+          // If this partition is not hosted by this host as per ideal state we stop tracking
+          removePartitionId(partitionId);
+          return;
+        } else {
+          // Partition is hosted by us as per ideal state, hence do not track for removal
+          // Reset timeout and try later if it continues to not consume.
+          _partitionsMarkedForConfirmation.put(partitionId, 0);
+        }
+      } else {
+        // If not enough time in inactive state, we just add to the timeout count
+        _partitionsMarkedForConfirmation.put(partitionId, currentTicks);
+      }
+    }
+  }
+
+  /*
+   * This function is invoked when a partition goes from CONSUMING to ONLINE, so we can assert whether the
+   * partition is still hosted by this server after some interval of time.
+   *
+   * @param partitionId Partition id that we need confirmed via ideal state as still hosted by this server.
+   */
+  public void markPartitionForConfirmation(long partitionId) {
+    _partitionsMarkedForConfirmation.put(partitionId, 0);
+  }
+
+  /*
+   * This is the function to be invoked when reading the metric.
+   * It reports the maximum Pinot Consumption delay for all partitions of this table being served
+   * by current server; it adds the time elapsed since the sample was taken to the measure.
+   * If no measures have been taken, then the reported value is zero.
+   */
+  public long getMaxPinotConsumptionDelay() {
+    DelayMeasure currentMaxDelay = getMaximumDelay();
+    if (currentMaxDelay == null) {
+      return 0; // return 0 when not initialized
+    }
+    // Add age of measure to the reported value
+    long measureAgeInMs = _enableAging ? (System.currentTimeMillis() - currentMaxDelay._sampleTime) : 0;
+    // Correct to zero for any time shifts due to NTP or time reset.
+    measureAgeInMs = measureAgeInMs >= 0 ? measureAgeInMs : 0;
```

Comment:
Again Math.max is more idiomatic and readable.

Reply:
agree, replaced over the change